src/pke/lib/scheme/bgvrns/bgvrns-parametergeneration.cpp

```diff
@@ -58,38 +58,187 @@ Archive, Report 2020/1118, 2020. https://eprint.iacr.org/2020/
+    std::static_pointer_cast<CryptoParametersBGVRNS>(cryptoParams);
+    //// HE Standards compliance logic/check
+  SecurityLevel stdLevel = cryptoParamsBGVRNS->GetStdLevel();
+
+  uint32_t ringDimension = cyclOrder / 2;
+
+  // RLWE security constraint
+  DistributionType distType =
+      (cryptoParamsBGVRNS->GetMode() == RLWE) ? HEStd_error : HEStd_ternary;
+  auto nRLWE = [&](usint q) -> uint32_t {
+    return StdLatticeParm::FindRingDim(distType, stdLevel, q);
+  };
+
+  // Case 1: SecurityLevel specified as HEStd_NotSet -> Do nothing
+  if (stdLevel != HEStd_NotSet) {
+    if (ringDimension == 0) {
+      // Case 2: SecurityLevel specified, but ring dimension not specified
+      // Choose ring dimension based on security standards
+      ringDimension = nRLWE(qBound);
+    } else {
+      // Case 3: Both SecurityLevel and ring dimension specified
+      // Check whether particular selection is standards-compliant
+      auto he_std_n = nRLWE(qBound);
+      if (he_std_n > ringDimension) {
+        OPENFHE_THROW(
+            math_error,
+            "The specified ring dimension (" + std::to_string(ringDimension) +
+                ") does not comply with HE standards recommendation (" +
+                std::to_string(he_std_n) + ").");
+      }
+    }
+  } else if (ringDimension == 0) {
+    OPENFHE_THROW(
+        math_error,
+        "Please specify the ring dimension or desired security level.");
+  }
+  return ringDimension;
+}
+
+std::pair<std::vector<NativeInteger>, uint32_t> ParameterGenerationBGVRNS::computeModuli(
+  std::shared_ptr<CryptoParametersBase<DCRTPoly>> cryptoParams,
+  uint32_t ringDimension,
+  int32_t evalAddCount, int32_t keySwitchCount,
+  usint relinWindow, usint firstModSize, uint32_t auxBits,
+  enum KeySwitchTechnique ksTech,
+  enum RescalingTechnique rsTech,
+  usint numPrimes) const {
+
+  uint32_t vecSize = (rsTech != FLEXIBLEAUTOEXT) ? numPrimes : numPrimes + 1;
+  std::vector<NativeInteger> moduliQ(vecSize);
+  
+  const auto cryptoParamsBGVRNS =
+      std::static_pointer_cast<CryptoParametersBGVRNS>(cryptoParams);
+  double sigma = cryptoParamsBGVRNS->GetDistributionParameter();
+  double alpha = cryptoParamsBGVRNS->GetAssuranceMeasure();
+  double plainModulus = static_cast<double>(cryptoParamsBGVRNS->GetPlaintextModulus());
+
+  // Bound of the Gaussian error polynomial
+  double Berr = sigma * sqrt(alpha);
+  // Bound of the key polynomial
+  double Bkey;
+
+  // supports both discrete Gaussian (RLWE) and ternary uniform distribution
+  // (OPTIMIZED) cases
+  if (cryptoParamsBGVRNS->GetMode() == RLWE) {
+    Bkey = sigma * sqrt(alpha);
+  } else {
+    Bkey = 1;
+  }
+
+  // delta
+  auto expansionFactor = 2. * sqrt(ringDimension);
+  // Vnorm
+  auto freshEncryptionNoise = Berr * (1. + 2. * expansionFactor * Bkey);
+  // V_c
+  auto noisePerLevel = 1 + expansionFactor * Bkey;
+
+  double keySwitchingNoise;
+  if (ksTech == BV) {
+    // TODO: Change tests to have this is a default for BV.
```

Comment:
@dsuponitskiy I have one unresolved issue here that I'm not sure how to solve. We're supposed to return an error if relinWindow = 0 for BV in BGV. But the default settings for tests has relinWindow = 0 and ksTechnique = HYBRID. I'm not sure how to set relinWindow = 4 as a default for BV.

Reply:
Fixed it by setting the default value in tests to 4.